Lib/Hash.hpp

```diff
@@ -116,156 +176,147 @@ class Hash
+  }
 
+  // overload for types with a content() method to get a hashable representation of the type
+  // e.g. TermList, SATLiteral
+  // hashes the result of content()
   template<typename T>
-  static unsigned hash(T o1)
-  { return Hash::hash(*o1); }
-};
+  static typename std::enable_if<
+    !is_safely_hashable<T>::value &&
+    is_safely_hashable<typename std::result_of<decltype(&T::content)(T)>::type>::value,
+    unsigned
+  >::type hash(const T &obj) {
+    return hash(obj.content());
+  }
 
-struct IdentityHash
-{
-  template<typename T>
-  static bool equals(T o1, T o2)
-  { return o1 == o2; }
+  // std::pair
+  template<typename T, typename U>
+  static unsigned hash(std::pair<T,U> obj)
+  { return HashUtils::combine(hash(obj.first), hash(obj.second)); }
 
-  template<typename T>
-  static unsigned hash(T val)
-  { return static_cast<unsigned>(val); }
-};
+  /**
+   * The FNV-hashing.
+   * @since 31/03/2006
+   */
+  static unsigned hash(const char* val) {
+    CALL("Hash::hash(const char *)");
 
-struct PtrIdentityHash
-{
-  static unsigned hash(const void* ptr) {
-    return static_cast<unsigned>(reinterpret_cast<size_t>(ptr));
+    unsigned hash = FNV32_OFFSET_BASIS;
+    while (*val) {
+      hash = (hash ^ *val) * FNV32_PRIME;
+      val++;
+    }
+    return hash;
   }
-};
 
-struct PtrPairSimpleHash {
-  template<typename T>
-  static unsigned hash(T pp) {
-    return static_cast<unsigned>(reinterpret_cast<size_t>(pp.first)^reinterpret_cast<size_t>(pp.second)^
-	    (reinterpret_cast<size_t>(pp.first)>>3)^(reinterpret_cast<size_t>(pp.second)>>4));
-  }
-};
+  /**
+   * The FNV-hashing.
+   * @since 31/03/2006
+   */
+  static unsigned hash(const unsigned char *val, size_t size)
+  { return hash(val, size, FNV32_OFFSET_BASIS); }
 
-struct IntPairSimpleHash {
-  template<typename T>
-  static unsigned hash(T pp) {
-    return static_cast<unsigned>(pp.first^pp.second^(pp.first<<1));
-  }
-};
+  /**
+   * The FNV-hashing where the initial value for hashing is @b hash.
+   * Useful for computing hash value of anything consisting of several
+   * parts: first the first part is hashed and then the hashing continues
+   * on the remaining parts but using the previously computed value as
+   * @b hash.
+   * @since 31/03/2006
+   */
+  static unsigned hash(const unsigned char *val, size_t size, unsigned hash) {
+    CALL("Hash::hash(const unsigned char *, size_t, unsigned)");
+    ASS(size > 0);
 
-template<class ElHash>
-struct ContainerHash {
-  template<typename T>
-  static unsigned hash(const T& cont) {
-    unsigned res = 2166136261u; //the FNV prime, don't know wheher it works well here:)
-    size_t sz = cont.size();
-    for(size_t i=0; i!=sz; i++) {
-      res = Hash::combineHashes(res, ElHash::hash(cont[i]));
+    for (int i = 0; i < size; i++) {
+      hash = (hash ^ val[i]) * FNV32_PRIME;
     }
-    return res;
+    return hash;
   }
 };
 
+// template specializations for SecondaryHash
```

Comment:
Lots of these are made considerably cheaper to compute, but worse as hashes. I don't see a problem with this given how secondary hashes are used.